### ISSUE TITLE
Regression in 1.70 - returns undef instead of time

### ISSUE THREAD

**Issue by TJC** [houseabsolute/DateTime-Format-Strptime#15]
See below script.
On 1.68, it returns `2016-12-14T23:37:12`, but on 1.70 it returns `undef`

```perl
#!/usr/bin/env perl
use 5.20.0;
use warnings;
use DateTime::Format::Strptime;

my $p = DateTime::Format::Strptime->new(
  pattern     => '%Y%m%d%H%M%S',
  locale      => 'en_US',
  time_zone   => 'UTC',
);

say $p->parse_datetime("20161214233712Z");
```

**Comment by autarch** [comment]
I agree that this is a regression but I'm torn on whether to fix it. This comes from a change in recent releases to make the date parsing look for word boundaries (regex `\b`) around the specified format.

In the example you gave we end up with a regex like `(?^:(?:\A|\b)(?^:((?^:(?^:(?:[0-9])){4})))(?^:((?^:[0-9 ]?(?^:(?:[0-9])))))(?^:((?^:[0-9 ]?(?^:(?:[0-9])))))(?^:((?^:[0-9 ]?(?^:(?:[0-9])))))(?^:((?^:[0-9 ]?(?^:(?:[0-9])))))(?^:((?^:[0-9 ]?(?^:(?:[0-9])))))(\b|\Z))`.

If we _don't_ include the `\b` bits then we could end up matching inside a larger number. In your example, we would basically match any number with 14+ digits, which can be really confusing. See #11 for another example of how not looking for word boundaries can lead to very confusing results.

I would note that in your case you could add a `%Z` to your pattern to make it work.

**Comment by TJC** [comment]
We managed to adjust our code to suit, but it took two developers around four hours each to locate the issue, as it was in some fairly old code that had "just worked" for quite a long time.

**Comment by autarch** [comment]
It occurs to me that maybe the best option would be to revert back to the old behavior and add a flag to enable the new, stricter behavior. Neither behavior is really likely to be right 100% of the time.

**Comment by preaction** [comment]
I think you're right about adding the flag: While I would expect people to carve the date/time string out of a larger string before trying to parse it with this module, they may not have been doing that. I don't think that the non-strict behavior is something one should be enthusiastic about, but requiring an opt-in after this long of a default is polite.

**Comment by autarch** [comment]
Fixed in 1.71
